compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEInsertDmaBdChain.cpp

```diff
@@ -0,0 +1,276 @@
+#define DEBUG_TYPE "iree-amdaie-insert-dma-bd-chain"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+using TileConnect = std::pair<AMDAIE::TileOp, AMDAIE::ConnectionOp>;
+
+/// Utility function to update `next_bd` and `start_bd` operands.
+LogicalResult updateChainOperands(
+    IRRewriter &rewriter, SmallVector<AMDAIE::NpuHalfDmaCpyNdOp> &dmaChain) {
+  // Nothing to do if the DMA chain length is one or less.
+  if (dmaChain.size() < 2) return success();
+
+  Value startBdId = dmaChain[0].getBdId();
+  Operation *parentOp = dmaChain[0]->getParentOp();
+  // Chain the DMA ops.
+  for (unsigned i = 0; i < dmaChain.size() - 1; ++i) {
+    AMDAIE::NpuHalfDmaCpyNdOp currDmaOp = dmaChain[i];
+    if (currDmaOp->getParentOp() != parentOp) {
+      return currDmaOp.emitError(
+          "DMA operations to be chained must belong to the same scope");
+    }
+    Value nextBdId = dmaChain[i + 1].getBdId();
+    // No token is produced at the beginning or middle of a chain.
+    TypeRange token = TypeRange{};
+    rewriter.setInsertionPointAfter(currDmaOp);
+    rewriter.create<AMDAIE::NpuHalfDmaCpyNdOp>(
+        currDmaOp.getLoc(), token, currDmaOp.getConnection(),
+        currDmaOp.getInput(), currDmaOp.getMixedOffsets(),
+        currDmaOp.getMixedSizes(), currDmaOp.getMixedStrides(),
+        currDmaOp.getBdId(), currDmaOp.getChannel(), nextBdId, startBdId);
+    for (auto &use : currDmaOp->getUses()) {
+      rewriter.eraseOp(use.getOwner());
+    }
+    rewriter.eraseOp(currDmaOp);
+  }
+  // Last DMA op in the chain.
+  AMDAIE::NpuHalfDmaCpyNdOp lastDmaOp = dmaChain.back();
+  if (lastDmaOp->getParentOp() != parentOp) {
+    return lastDmaOp.emitError(
+        "DMA operations to be chained must belong to the same scope");
+  }
+  Value nextBdId = nullptr;
+  rewriter.setInsertionPointAfter(lastDmaOp);
+  auto lastDmaOpChained = rewriter.create<AMDAIE::NpuHalfDmaCpyNdOp>(
+      lastDmaOp.getLoc(), lastDmaOp.getResultTypes(), lastDmaOp.getConnection(),
+      lastDmaOp.getInput(), lastDmaOp.getMixedOffsets(),
+      lastDmaOp.getMixedSizes(), lastDmaOp.getMixedStrides(),
+      lastDmaOp.getBdId(), lastDmaOp.getChannel(), nextBdId, startBdId);
+  rewriter.replaceOp(lastDmaOp, lastDmaOpChained.getResults());
+  return success();
+}
+
+/// Utility function to determine if chains can grow further
+/// or require breaking.
+///
+/// Example:
+/// - Chain X currently holds BD IDs: [4, 5, 6, 7]
+/// - Chain Y currently holds BD IDs: [0, 1, 2, 3]
+/// - A new BD ID (0) needs to be added to the front (due to reverse
+/// traversing) of chain X.
+///
+/// Conflict resolution:
+/// - Chain Y must be broken because BD ID 0 is already assigned to it
+/// and must be released.
+/// - Chain X is also broken to prevent the new added BD ID (0) from
+/// invalidating chain Y.
+///
+/// Result:
+/// - Break both chains X and Y.
+///   - Chain X: [0] (the newly added BD ID).
+///   - Chain Y: [] (emptied after breaking).
+void canChainGrowFurther(
```

Comment:
Someone reading `canChainGrowFurther` would expect a bool to be returned, but this is actually find and updating chains to be broken, so I think this should have a more descriptive name, like `checkForChainsToBeBroken`.